### DISCUSSION TITLE
Doubt about security of scvelo results.

### DISCUSSION THREAD

**Discussion by alexmascension** [theislab/scvelo#393] [Q&A] [answered]
Hi! I've been using scvelo recently for some analysis, and I've come across some results that we striking to me.

My data are some cells at days 0, 2, and 4, and I've run the analysis as in the mock data from this notebook
https://colab.research.google.com/drive/1I-KV-QSJU7v1GuzIwzI9lSY2CiAAIKj7?usp=sharing

The point is that when I did the scatter of several genes (marker genes selected within the heatmap) against the latent time, I realized that the Ms values were the ones plotted by default in the scatter plot, and were are *far off* from the expression values, in X, spliced, and unspliced layers. In general, these many genes are expressed in few cells whereas the Ms values are not zero in most of the cells. Despite that, I see a slight correlation between the Ms and expression values.

![image](https://user-images.githubusercontent.com/35657291/110994755-90e2db80-8379-11eb-9203-8230c7dff9e7.png)
![image](https://user-images.githubusercontent.com/35657291/110995482-b7eddd00-837a-11eb-94b7-b95d48afb03c.png)
![image](https://user-images.githubusercontent.com/35657291/110995550-d2c05180-837a-11eb-987d-5c49f2c286ea.png)
![image](https://user-images.githubusercontent.com/35657291/110995701-10bd7580-837b-11eb-95d1-d5dc958f4dc0.png)

My doubt is to which extent can I *believe* the predicted Ms values as a proxy for a gene that is up/downregulated across days. This dataset only has ~150 cells and, if the expressed cells are ~20 (although most of the cells show positive Ms values), can I believe the analysis to infer biological hypotheses? I tried to read your paper and I have doubts about how the Ms is recovered.

Thanks for your help!

**Comment by WeilerP** [comment, marked as answer]
First order moments are defined as the mean expression of the _k_ nearest neighbors of a cell. Not sure what you mean by "predicted Ms" or "recovering" them as moments are not a real physical variable you can measure (and therefore not recover or predict). The moments are used to smoothen measurements and reduce the sparsity of the data. This step is crucial for even having a chance to observe splicing dynamics.

Sorry, not sure what you are plotting in the given figures and why it is helpful. Also, what do you mean by "expressed cells"?

**Comment by alexmascension** [comment]
Hi! Thanks for the help.

I understand the Ms now. What I show in the pictures is the scatter plot of each of the variables produced by scvelo, because I didn't know exactly what was being plotted by the scatter function with the default settings (first subplot). By "expressed cells" I mean cells with expression > 0. 

The point is, although I get that using kNN expression is necessary, I have doubts about to what extent can I directly believe the analysis because with my data I find examples where the gene expression is extremely low and, nonetheless, the Ms values seem more "realistic".  

For example, some top_genes in my analysis are IGF1R, BMP2 and DPP4:

![image](https://user-images.githubusercontent.com/35657291/111336005-21b90000-8675-11eb-8a00-87690fb55739.png)
![image](https://user-images.githubusercontent.com/35657291/111336053-2d0c2b80-8675-11eb-9a0e-7bdeda1baf8b.png)
![image](https://user-images.githubusercontent.com/35657291/111336135-401efb80-8675-11eb-89fa-745d2b1de7c5.png)

You can see that, for example, BMP2 is only expressed in 3 cells out of the 122, but due to the kNN, it looks like many more cells are expressed and, coincidentally, there's no expression at day 0. It is true that, for some genes with more expression, the smoothing makes the Ms scatter more appealing, and it is more trustworthy, but many of these low-expression genes are not filtered and can look like artifacts IMO. 

If I want to retain fewer genes but with more expression and whose Ms I can trust, what should I do? Should I reduce k?

Thanks!

**Comment by WeilerP** [comment]
Okay, so there are only ~20 cells containing expressed genes? Not sure if I understood correctly, but cells containing only zero counts should be removed during quality control.
Also, as part of the preprocessing, genes expressed in only a small number of cells should be excluded from any downstream analysis as well. You can do so via `scv.pp.filter_genes`. Alternatively, you can also use `scv.pp.filter_and_normalize` which simply combines filtering and normalizing.

**Reply by alexmascension** [reply]
Okay, thanks. 

No, in this case I mean that this gene is expressed in 20 cells. In that case, the gene might be filtered out because it has around 10 reads (without normalization + log1p), but still, I am interested in knowing if it would be feasible to lower the k value.

**Reply by WeilerP** [reply]
Okay, gotcha. Just FYI, if you want to prevent certain genes from being filtered, you can specify them using the keyword argument `retain_genes`.
Generally speaking, I see one potential problem arising from the overall low number of cells in your dataset: Choosing _k_ large (w.r.t. the overall number of cells), cells considered to be neighbors may not actually be very similar. So choosing a smaller _k_ can be feasible. However, there is always a trade-off since lowering _k_ will also increase sparsity and potentially noise.

**Reply by alexmascension** [reply]
Okay. I will try with a lower k value at that expense. Thanks for your help! 😊
